Adds the ability to store knives/pens/edaggers in certain shoes

PKPenguin321 opened the pull request:
Stored by clicking the shoe with said items, and removed by click-dragging the shoe with the item to yourself/an empty hand slot.
Placing an item into the shoe shows a message to only the user. Taking an item out, however, warns everybody in the vicinity.

:cl: PKPenguin321
rscadd: You can now store knives, pens, switchblades, and energy daggers in certain types of shoes (such as jackboots, workboots, winter boots, combat boots, or no-slips).
rscadd: Horns can now be stored in clown shoes. Honk.
/:cl:

PKPenguin321 committed:
adds the ability to store knives/pens/edaggers in certain shoes
added by clicking the shoe with said items, and removed by click-dragging the shoe with the item to yourself/an empty hand slot

Zilenan91 commented:
Can you make it so the clown can store his bike horn in his shoes

PKPenguin321 commented:
hypothetically yes but it would be really snowflake and too much work to be worth doing as far as i can tell

Zilenan91 commented:
Isn't this whole thing already snowflake? How different would it be to add the bike horn to that?

PKPenguin321 commented:
this whole thing is fairly generalized by being a var placed on the parent shoe type that, when set to 1 on a child, allows the child shoe type to store knives. the hackiest thing about it is either the way it puts the knife in your hand when you click-drag to yourself or the double istype check.

an extra check for specifically clown shoes for specifically bike horns would be pretty snowflake since it would break the generalized system i worked so hard to establish and essentially require me to copy-paste the entire system for knife-storing to clown shoes just so i could change one line to be "horn" instead of "knife."

Zilenan91 commented:
Or you could just make bike horns in general be able to be stored in all shoes rather than just clown shoes

PKPenguin321 commented:
i *guess* but it feels like that would detract from the clownyness of it

Zilenan91 commented:
In practice only clowns would make use of this feature.

Supermichael777 commented:
too bad pat-down checkpoints are so rare this might be cooler if they were.
can you get it out shoe inhand?

PKPenguin321 commented:
yeah, just have the shoe in your hand and drag from the shoe to yourself/your other hand

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -175,6 +175,8 @@ BLIND     // can't see anything
 	slowdown = SHOES_SLOWDOWN
 	var/blood_state = BLOOD_STATE_NOT_BLOODY
 	var/list/bloody_shoes = list(BLOOD_STATE_HUMAN = 0,BLOOD_STATE_XENO = 0, BLOOD_STATE_OIL = 0, BLOOD_STATE_NOT_BLOODY = 0)
+	var/can_hold_knives = 0//if set to 1, the shoe can hold knives and edaggers
```
This is really snowflake-y...

PKPenguin321 replied:
not really, it's just an object var on a parent type object. same type of stuff is all over the place, like on energy guns self-charge var

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,46 @@ BLIND     // can't see anything
+		M.put_in_hands(held_knife)
+		usr.visible_message("<span class='warning'>[usr] draws [held_knife] from \his shoes!</span>", "<span class='notice'>You withdraw [held_knife] from [src].</span>")
+		held_knife = null
+	else if(istype(over_object, /obj/screen))
```
Fairly sure we have `put_in_hands` for this.

PKPenguin321 replied:
that's used a few lines up for when it's dragged to your character itself. this is used for specifically dragging to your left/right hands. it's kind of weird, but it works.

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,46 @@ BLIND     // can't see anything
+		switch(over_object.name)
+			if("r_hand")
+				if(!remove_item_from_storage(M))
+					M.unEquip(held_knife)
```
Don't think this is needed if its held inside the contents of another object?

PKPenguin321 replied:
as for this i honestly have no clue since i basically copied down paper bin's code for dragging things into your hands. as far as i know it works as intended, and i worry if i remove this something will break, but if you're sure it's not need i'll remove it.

neersighted replied:
This code is just obviously copypasta and a bit... wrong... Please drop by coderbus and have someone more knowledgeable (I'm not totally up to date on inventory handling) walk you through it.

PKPenguin321 commented:
probably gonna change click-dragging to alt-clicking soon^TM

WJohn commented:
Click dragging to remove an accessory is a pretty terribly inconsistent thing with everything else we have. I don't think jumpsuit accessories are even removed that way. Since jumpsuits are capable of showing their accessories as little shrunken icons in the top right, why not make it so you can click on that icon directly to remove it?

Zilenan91 commented:
Jumpsuits use the right click system. You could probably use that I guess if you handle the knives/bike horns as accessories on the shoe.

PKPenguin321 commented:
nope nope nope
object verbs are a no-no
don't make me call @MrPerson on you guys for all this blasphemy

PKPenguin321 committed:
makes drawing the knife use alt click instead of click drag
so many saved lines
oh yeahhh feels good
thanks @neersighted

Anonmare commented:
Finally this bug is being fixed

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,28 @@ BLIND     // can't see anything
+			return
+		I.loc = src
+		user << "<span class='notice'>You discreetly slip [I] into [src]. Alt-click [src] to remove it.</span>"
+		held_knife = I
```
message should come after `held_knife`.

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,28 @@ BLIND     // can't see anything
+	if(!user.put_in_hands(held_knife))
+		user << "<span class='notice'>You fumble for [held_knife] and it falls on the floor.</span>"
+		return 1
+	user.visible_message("<span class='warning'>[user] draws [held_knife] from their shoes!</span>", "<span class='notice'>You withdraw [held_knife] from [src].</span>")
```
Just draw, not withdraw...

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,28 @@ BLIND     // can't see anything
+		return 1
+	user.visible_message("<span class='warning'>[user] draws [held_knife] from their shoes!</span>", "<span class='notice'>You withdraw [held_knife] from [src].</span>")
+	held_knife = null
+	add_fingerprint(user)
```
Fingerprint should probably go on the knife.

PKPenguin321 replied:
actually i dont think this line is needed at all on second thought, you get fingerprints on the knife from holding it already and to put shoes on your feet covers them in fingerprints to begin with. i'll just delete this line

neersighted commented on code/modules/clothing/shoes/miscellaneous.dm:
```diff
@@ -16,6 +16,7 @@
 	flags = NOSLIP
 	origin_tech = "syndicate=3"
 	burn_state = FIRE_PROOF
+	can_hold_knives = 1//syndicate-brand shoes come with syndicate-brand knife departments
```
compartments :p

PKPenguin321 replied:
would fix but there's no point it's just a comment :stuck_out_tongue:

tkdrg replied:
no comments > bad comments
improve or remove

LanCartwright commented:
Interesting story. I was discussing on coderbus about the inclusion of weapons attachable to jumpsuits and they hated the idea, instead referring me to try and put items into shoes.

Robertha89 commented:
Why not making a whitelist/blacklist for what you can/can't put in the shoes instead?

neersighted commented:
Because whitelists and blacklists are bad.

Robertha89 commented:
Are there any other ways to make it a modular system where items can be added or removed instead of wl and bl

PKPenguin321 commented:
a big nasty istype

PKPenguin321 committed:
addresses some various feedback

changes position of a message, tweaks a message, removes useless fingerprint line

neersighted commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,27 @@ BLIND     // can't see anything
+	if(held_knife)
+		user << "<span class='notice'>There's already something in [src].</span>"
+		return
+	if(istype(I, /obj/item/weapon/kitchen/knife) || istype(I, /obj/item/weapon/pen))//can hold both regular pens and energy daggers. made for your every-day tactical librarians/murderers.
```
This really could be generalized to just include very small items.

PKPenguin321 replied:
i considered it but i figured that would be too much. it would be like having a third pocket that can't be pickpocketed and this feature isn't supposed to be THAT powerful.

neersighted replied:
This istype just bothers me too much... Shoes that can hold stuff are rarer so it seems fair enough to allow something tiny to be stashed in there. Also keeps you from putting a butchers cleaver in your boot (ouch!).

PKPenguin321 replied:
yeah the cleaver thing is kinda funny but i let it slide
jackboots/workboots/snowboots are not rare though

PKPenguin321 committed:
fixes a case where held_knife wouldn't null after knife was removed

O N E   L I N E   C O M M I T S
N
E

L
I
N
E

C
O
M
M
I
T
S

neersighted commented:
:+1: looks good to me.

neersighted commented:
:+1: looks good to me.

PKPenguin321 committed:
allows switchblades too

still no bike horns, rip zilenan

tkdrg commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,28 @@ BLIND     // can't see anything
+			return
+		I.loc = src
+		held_knife = I
+		user << "<span class='notice'>You discreetly slip [I] into [src]. Alt-click [src] to remove it.</span>"
```
It would be nice if this proc could call parent somewhere, for future-proofing. (Also would be nice if you did the same for banana_shoes attackby() if possible)

PKPenguin321 replied:
im not 100% what you mean, you mean move all this to the parent of shoes (which would be item/clothing)..? also i'm not sure how i would incorporate the banana shoes attackby() into this at all...
ill ask around for you in coderbus but if we miss each other pls resond here

tkdrg commented on code/modules/clothing/shoes/miscellaneous.dm:
```diff
@@ -115,6 +119,7 @@
 	item_state = "jackboots"
 	strip_delay = 40
 	put_on_delay = 40
+	can_hold_knives = 1//for when that wire is too tough for regular wirecutters
```
this one doesn't make sense to me. also no meme comments pls

PKPenguin321 replied:
:(

PKPenguin321 committed:
removes quality humor

comments are srs bisnis

tkdrg commented:
You forgot to call parent though

PKPenguin321 committed:
calls my parents

they were getting lonely

ps i actually have no idea what this does it might break things

tkdrg commented on code/modules/clothing/clothing.dm:
```diff
@@ -198,6 +200,29 @@ BLIND     // can't see anything
+	if(held_knife)
+		user << "<span class='notice'>There's already something in [src].</span>"
+		return
+	if(istype(I, /obj/item/weapon/kitchen/knife) || istype(I, /obj/item/weapon/pen) || istype(I, /obj/item/weapon/switchblade))//can hold both regular pens and energy daggers. made for your every-day tactical librarians/murderers.
```
This list could probably be a var in shoes instead. That way you could use is_type_in_list() here and have different shoes keep different whitelists

PKPenguin321 replied:
as far as i can tell this is already a var in shoes
is_type_in_list sounds cool though, educate me

tkdrg replied:
it's not a var it's a hardcoded check
make it a var/list in the shoes and is_type_in_list I on it

PKPenguin321 replied:
oh, you mean make the thing the istype() checks for itself into a var?

tkdrg replied:
Make this a shoe var
`var/list/can_be_hidden = list(obj/item/weapon/kitchen/knife, /obj/item/weapon/pen, /obj/item/weapon/switchblade)`
Then replace this line with
`if(is_type_in_list(I, can_be_hidden))`
That way the hidden list is more flexible and can be overriden by shoe children/creative admins.

PKPenguin321 replied:
[you smart](https://www.youtube.com/watch?v=Ll9nfNlDLII)

PKPenguin321 committed:
changes istype to is_type_in_list, adds a w_class check

also changes some var names to be less knife-specific and more item-specific

PKPenguin321 committed:
updates var names, allows clown shoes to hold bike horns

zilenans around the globe, rejoice

PKPenguin321 committed:
fixes travis

o n e    l i n e    c o m m i t s
